### ISSUE TITLE
BigTreetech TFT 24 not working

### ISSUE THREAD

**Issue by erik23de** [KevinOConnor/klipper#2231]
Hi,

Tried to set up Klippy on an skr 1.3 (from the example- configs), copy & pasted the Preprap GLCD
into the config.
Usually, testet with 3 Boards, Marlin and Repetier, the BigTreetech TFT 24 (and 35) works as a 12864 GLDC like a charm (First Display I'm really happy with).
But it does not with Klipper, Screen stays empty (but reset switch works). Wiring is okay, it works with Marlin on the same board.
What did I wrong?

[klippy.zip](https://github.com/KevinOConnor/klipper/files/3908316/klippy.zip)

**Comment by KevinOConnor** [comment]
I don't know of anyone that has attempted to use this device with Klipper.  It appears the device is trying to emulate an st7920 device.  As a guess, it may not fully emulate the device and thus not be compatible with Klipper.  (Klipper uses some features of the st7920 - such as the character generator - that other firmware don't use.)

-Kevin

**Comment by erik23de** [comment]
Thanks for the info.

You should give it a chance - it's a good & cheap Display and we'll see more of that around soon.
With more conventional firmware, it works quiete easy with the widespread RepRapDiscount GLCD configuration.

In meantime, maybe I'm luckier with Klipper and the fysetc display - with wich I had no chance on Marlin or Repetier.

**Comment by jalanjarosz** [comment]
@erik23de   I have a Fysetc MKS Mini display with the NeoPixel LED's working though a RAMPS shield on an Adafruit Grand Central M4.  Let me know if you need a hand with the config.

**Comment by RomRider** [comment]
Same here with TFT35, if you switch back and forth between the touch mode and the marlin mode, it will ultimately display a bunch of garbage after playing with the button:
![image](https://user-images.githubusercontent.com/21064206/70312664-60c29980-1814-11ea-83c3-e9008cdaa043.png)

**Comment by derBG** [comment]
Iam also interested. Does anyone have a config that shows anything? My TFT24 connected to skr 1.3 shows in the 12864 mode only a blue display.

**Comment by matty-byte** [comment]
I have a TFT24, neither the emulated screen nor the touch screen will connect. The screen defaults to 250k baud rate which does not connect. Changing the baud rate to 115k does nothing and when rebooting it resets to 250k. BTT have guides for Marlin, but I have not been able to work out if any of these settings can be configured with Klipper.

**Comment by KevinOConnor** [comment]
FYI, it appears that the tft24 (and similar) are emulating an st7920 chip:

https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/blob/e180179aa945cfcf4a5af5c71581138d4d87225d/TFT/src/User/API/UI/ST7920_Simulator.c

That emulation does not include the character generator, so it is not compatible with Klipper.  However, if a developer is interested, it should be possible to add support to Klipper.

-Kevin

**Comment by softkannan** [comment]
1.	BTT and MKS TFT both emulates host (like octoprint), no config. The incoming serial data is  GCODE commands and they listen for Ok and other serial command just like octoprint. Marlin tackled these displays by having more than one serial port can act as host.

2.	Anycubic TFT display also sends serial texts on high level but key difference is Anycubic TFT is not acting like host, its serial commands needs to be translated (requires intermediator).

3.	All the above 3 displays offer no config, means we cannot add new buttons or any other menus without compiling display firmware.

4.	to support above type of displays, on klipper we need
      **on mcu side**
       1. Generic serial text send
       2. Generic serial  text recev
      **on klippy side**
       1. Generic gcode injector to queue and response, it has to emulate klippy virtual serial port functions, like sending ok, busy etc. 
       2. Once above “Generic gcode injector to queue” is ready then we can customize for Anycubic TFT type displays.

**Benefit**
        - It makes klipper deployment easy, no hardware modification required, except plugging in rpi. 
        - Above approach is not specific to any display (except anycubic, I believe that must be done on private), which also preserve future expandability.
**Cons**
	It might increase some traffic, but I guess impact on performance is minimal.

Is above approach make sense ?

**Comment by alagerqvist** [comment]
What klipper display settings are you using while trying to get the TFT24/35 running.

This is my interpretation of how to set it up.
[display]
lcd_type: st7920
cs_pin: PD10
sclk_pin: PG2
sid_pin: PD11
encoder_pins: PG10,PF11
click_pin: PA8
# Klipper     Marlin
# cs_pin == LCD_PINS_RS
# sclk_pin == LCD_PINS_D4
# sid_pin == LCD_PINS_ENABLE
# encoder_pins == BTN_EN1, BTN_EN2
# click_pin == BTN_ENC
[output_pin beeper]
d4_pin: PG2
d5_pin: PG3
d6_pin: PG6
d7_pin: PG7

[output_pin beeper]
pin:  PG4

[gcode_macro M300]
default_parameter_S=1000
default_parameter_P=100
gcode:  SET_PIN PIN=beeper VALUE=1
        G4 P{P}
        SET_PIN PIN=beeper VALUE=0

Tested the M300 to get a beep from the display =D

**Comment by DeviousPenguin** [comment]
@alagerqvist I agree serial communication with the screen in 'touch' mode would be nice, I'm currently experimenting with using a USB2TTL adaptor connected from the RPi directly to the Display, and while it connects, I'm unable to get it to do anything useful so far.

This way the USB link betweek Klipper and the MCU is totally unhindered, the Screen is on a different bus, and can even be a different baud rate, but the wiring for this is less clean if your RPi is not mounted to your printer.

Can I ask if you managed to get the 'marlin mode'/'12864 simulator' mode to get anything to appear on the screen? Or is it only the Beep that works?

**Comment by alagerqvist** [comment]
@DeviousPenguin  
I have tried a bit of debugging but my problem is that I don't receive ANY serial communication from klipper, it is probably due to my display set-up in klipper, but the problem is that the documentation for setting up display for klipper is close to none.
In my previous comment I tried to describe my guess on how to map Marlin configurations to klipper as marlin configurations are easy to come by for any hardware.

So if someone could confirm/oppose the below I'd be happy: (Do I lack something?)
Klipper <==> Marlin:
cs_pin <==> LCD_PINS_RS
sclk_pin <==> LCD_PINS_D4
sid_pin <==> LCD_PINS_ENABLE
encoder_pins <==> BTN_EN1, BTN_EN2
click_pin <==> BTN_ENC
 
I have started on an implementation of "full 7920 display emulation" for the TFT24/35 (or other), but I really can't test until I get serial working against klipper.
It look quite straightforward to change the 7920 code in klipper but I'm not sure that is the way to go... (but serial multiplexing is a must have....)

**Comment by DeviousPenguin** [comment]
@alagerqvist I may be wrong but 7920 display emulation mode doesn't need serial in order to work.

I had a SKR v1.4 board turn up today, once I get it plugged up and flashed with Klipper, I'll see if I can confirm your pinouts and get some form of functionality with the TFT24

**Comment by alagerqvist** [comment]
Hi,

I was a bit unclear, when I said serial was'nt working I actually meant the
SPI communication to the display.
So what I mean is that the screen gets no 7920 instructions over the
configured pins, I added on screen debug info to see.

Den lör 29 feb. 2020 03:00DeviousPenguin <notifications@github.com> skrev:

**Comment by alagerqvist** [comment]
@DeviousPenguin 
I was a bit unclear, when I said serial wasn't working I actually meant the SPI communication to the display.  So what I mean is that the screen gets no 7920 instructions over the configured pins, I added on screen debug info to see if the screen get any data. 

The below might have some impact: (This is the setting that works against the display using Marlin)
    #define BOARD_ST7920_DELAY_1 DELAY_NS(96)
    #define BOARD_ST7920_DELAY_2 DELAY_NS(48)
    #define BOARD_ST7920_DELAY_3 DELAY_NS(600)
I have no idea if possible/how to set those delays for Klipper...

**Comment by alagerqvist** [comment]
@KevinOConnor 
I took a look at the low-level specification for communicating with the 7920 and what Klipper does seems a bit of the spec for TS7920.
Setting the clock-pin to low before setting the data and then pulling clock up seems correct according to the data sheet. (So data is read on rising clk).
(According to the spec. page 38 "Serial interface timing diagram  MPU write data to ST7920")

The timing of Klipper looks strange and the clk is toggled instead of set.

Marlin does this: (Works with BTT code which isn't real 7920 HW) 
  WRITE(ST7920_CLK_PIN, LOW);             ST7920_DELAY_1; \
  WRITE(ST7920_DAT_PIN, ST7920_DAT(val)); ST7920_DELAY_2; \
  WRITE(ST7920_CLK_PIN, HIGH);            ST7920_DELAY_3; \

Klipper does:
        if (data & 0x80) {
            gpio_out_toggle(sid);
            data = ~data;
        }
        ndelay(200);
        gpio_out_toggle(sclk);
        ndelay(200);
        data <<= 1;
        gpio_out_toggle(sclk);

**Comment by KevinOConnor** [comment]
I'm not aware of any issues with the Klipper bit banging interface to the st7920.  What you're looking at is the low-level toggling that is done - it's different from Marlin, but I don't believe it is incorrect.

Separately, when I last looked at the BTT firmware running on the display (https://github.com/KevinOConnor/klipper/issues/2231#issuecomment-566371883) it did not support the st7920 commands that Klipper issues.  I'm pretty sure that's the root of the issue.

-Kevin

**Comment by alagerqvist** [comment]
Hi,

I was actually checking against the hardware sheet, I understand that
Marlin isn't a golden standard of implementation for this.

Anyway, I have attached a picture from the specification showing what I
mean, there is also a timing table that is dependant on the Vcc 3.3V or 5V
for the TS7920.

My point is that trying to follow the hardware spec. seems to make the
display software fail for Klipper.

Cheers/
Anders Lagerqvist


Den mån 2 mars 2020 04:20KevinOConnor <notifications@github.com> skrev:

**Comment by 4noxx** [comment]
hi, are there a solution to run the bigtreetech tft-display (perhaps in marlin mode)
in klipper?
i use a SKR E3 MINI, and no Display runs on my Board (test bigtreetech tft35 and an reprap graph display and my old anet a8 display).

**Comment by JohnEdwa** [comment]
AFAIK, no. Klipper requires the display to support the character generator, which the BTT emulator doesn't have. There is [an open issue on the firmware page](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/issues/290) about it, but nobody has implemented it yet.   
Also, as it's not needed by Marlin, only by Klipper and other misc projects, the chances are unless someone from here goes to implement it, it never will.

**Comment by DeviousPenguin** [comment]
> Also, as it's not needed by Marlin, only by Klipper and other misc projects, the chances are unless someone from here goes to implement it, it never will.

I'm not the most experienced with C++ and Arduino type code, but can do the basics. Would you know the best place to start with implementing the character generator?

**Comment by trevjonez** [comment]
> > Also, as it's not needed by Marlin, only by Klipper and other misc projects, the chances are unless someone from here goes to implement it, it never will.
> 
> I'm not the most experienced with C++ and Arduino type code, but can do the basics. Would you know the best place to start with implementing the character generator?

From the other thread just recently:
https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/issues/290#issuecomment-610286699

It sounds like they may not emulate the char generator, but use font files that are already present in the firmware.

**Comment by DeviousPenguin** [comment]
@trevjonez thanks it does look like BTT may be adding the char gen, by the looks of that thread, so we may not need to re-invent the wheel

**Comment by dushyantahuja** [comment]
I've just got this TFT (a few minutes back), and see that there's a USB port behind it. Would it be possible to connect this to the pi and simply link the two serial port together (one from the TFT and /tmp/printer) - so that any commands the touch screen sends would get sent to klipper.

I'm no Linux expert - but it might be possible using socat. Will have to try once I get my board installed and have the printer running again. 

http://technostuff.blogspot.com/2008/10/some-useful-socat-commands.html

**Comment by JohnEdwa** [comment]
That port is only for connecting a USB flash drive to the printer instead of using the SD card slot. The serial port is on a different connection entirely, as that is indeed how the display communicates using the "touch screen mode" with the mainboard: purely by sending and receiving Gcode commands, just like OctoPrint does.

Connect that display to one of the serial ports on your Raspi instead of the mainboard and instruct Klipper to listen to that port, and you could, in theory, replace Octoprint with that touchscreen. With the caveat that it's almost guaranteed to use some Marlin specific commands that won't work with Klipper.

**Comment by dushyantahuja** [comment]
Aah - too bad. Turned out when I shorted my previous control board, it also took the TMC drivers with it :(. Now to wait for the new ones to arrive, before I can experiment further. More thoughts:

1. As Kevin mentioned, port Klipper directly to TFT and use it as another microcontroller. Since I have no experience in this, I'll leave it to the experts
2. Connect to ESP8266 - and then use the Octoprint API to get / set data. It shouldn't be too difficult, let me see what I can do.

**Comment by HenrikPPersson** [comment]
Hi! New owner of Bigtreetech TFT24 and first poster on Github.
I looked through the source of the TFT24 firmware and discovered that the display only emulate the ST7920 graphics mode. Not the text or glyph mode. After modifying the st7920.py in klipper I sort of got it working.

When I start up the screen, it is blank, part from the "LCD12864 Simulator" text at the top. Once I switch over to Serial Touch Screen mode and then back again it starts to show things. With the patch below, it seems to be working (part from the animations).

```diff
@@ -126,7 +126,11 @@ class ST7920:
         if x + len(data) > 16:
             data = data[:16 - min(x, 16)]
         pos = [0, 32, 16, 48][y] + x
-        self.text_framebuffer[pos:pos+len(data)] = data
+        # self.text_framebuffer[pos:pos+len(data)] = data
+        for c in data:
+            for i, bits in enumerate(bytearray(font8x14.VGA_FONT[ord(c)])):
+                self.write_graphics(x, y, i, [bits])
+            x=x+1
     def write_graphics(self, x, y, row, data):
         if x + len(data) > 16:
             data = data[:16 - min(x, 16)]
```

**Comment by trevjonez** [comment]
what kind of increased load does that put on the MCU? seems like a decent short term workaround while we wait for btt to finish the char generator.

**Comment by HenrikPPersson** [comment]
So far I haven't run it more than a couple of hours, so long time effects are unknown. The solution for uc1701 is very similar to this, so the the MCU load shouldn't be much worse than with that one.

Right now, I have it hooked up to a spare controller board (Trigorilla) and running klipper directly on my development WS (Fedora). I will connect it to one of my printers and see how it behaves in real life.

**Comment by nklrk** [comment]
FYI: Yesterday (23.05.2020) BTT added support for klipper (https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/commit/e6d8218dc20d4d5d4b3a2700c0e3fc26bb4059ed)

**Comment by alagerqvist** [comment]
Hi,

Sorry, I haven't tried for a long while.
I think the low level bit banging timing is the root problem then the
character emulator needs to be implemented.

/Anders

Den ons 26 feb. 2020 01:44DeviousPenguin <notifications@github.com> skrev:

**Comment by dushyantahuja** [comment]
Hmmm - just uploaded the latest firmware to my TFT24 (compiled from the latest source) - and have the following in my config:

```
[board_pins]
aliases:
    # EXP1 header
    EXP1_1=P1.30, EXP1_3=P1.18, EXP1_5=P1.20, EXP1_7=P1.22, EXP1_9=<GND>,
    EXP1_2=P0.28, EXP1_4=P1.19, EXP1_6=P1.21, EXP1_8=P1.23, EXP1_10=<5V>,
    # EXP2 header
    EXP2_1=P0.17, EXP2_3=P3.26, EXP2_5=P3.25, EXP2_7=P1.31, EXP2_9=<GND>,
    EXP2_2=P0.15, EXP2_4=P0.16, EXP2_6=P0.18, EXP2_8=<RST>, EXP2_10=<NC>
    # Pins EXP2_1, EXP2_6, EXP2_2 are also MISO, MOSI, SCK of bus "ssp0"

[display]
lcd_type: st7920
cs_pin: EXP1_4
sclk_pin: EXP1_5
sid_pin: EXP1_3
encoder_pins: ^EXP2_3, ^EXP2_5
click_pin: ^!EXP1_2
#kill_pin: ^!EXP2_8
```
Still doesn't work - I just get a blank screen with "LCD12864 simulator" on top
Has anyone got it working - can you please advise if I've got the pins wrong, or maybe something else.

**Comment by HenrikPPersson** [comment]
There seems to be something wrong with the initialization of the ST7920 mode in the TFT24-display. I have to switch over to Touch mode and then back to Marlin mode after i power on. Otherwise the display stays black.

**Comment by dushyantahuja** [comment]
Got it working - needed to change
```
SPI_ReEnable(!!(GPIOB->IDR & (1<<13))); to SPI_ReEnable(1)
```
in https://github.com/Msq001/BIGTREETECH-TouchScreenFirmware/blob/236c497d6050215b9674f6f78ffe0d581a834417/TFT/src/User/Hal/stm32f10x/spi_slave.c#L119

There's also a binary in the pull request that you can use directly:
https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/pull/705

**Comment by xkuyax** [comment]
Hi @dushyantahuja thanks for you work on this! 
Could you help me to get my LCD working?

I have the following setup:
Klippy v0.8.0-526-gf0c394de
Skr Pro 1.1 
TFT35 V3 with the newest Firmware (cloned fresh from git) and i did the change in spi_slave 

At first i attached the TFT Dupont Cable, but i disconnected that (as it interfered with klipper)

Now i only have exp1 and exp2 connected and the following in my config:
[board_pins]
aliases = 
	EXP1_1=PG4, EXP1_3=PD11, EXP1_5=PG2, EXP1_7=PG6, EXP1_9=<GND>,
	EXP1_2=PA8, EXP1_4=PD10, EXP1_6=PG3, EXP1_8=PG7, EXP1_10=<5V>,
	EXP2_1=PB14, EXP2_3=PG10, EXP2_5=PF11, EXP2_7=PF12,  EXP2_9=<GND>,
	EXP2_2=PB13, EXP2_4=PB12, EXP2_6=PB15, EXP2_8=<RST>, EXP2_10=PF13
(taken from the example of the skr pro)

[display]
lcd_type = st7920
cs_pin = EXP1_4
sclk_pin = EXP1_5
sid_pin = EXP1_3
encoder_pins = ^EXP2_3, ^EXP2_5
click_pin = ^!EXP1_2

But i only see the LCD12864 when selecting the marlin mode...
Anything special i need to configure in the config.ini or something else?

**Comment by dushyantahuja** [comment]
The firmware is not perfect yet. Do you have a knob on your TFT? Sometimes
I see the blank screen as well, but if I press the knob a couple of times,
the menu comes up.

Try changing to touch screen mode and then back to classic mode. That
should work.

Last option would be to compare the pins on the SKR 1.1 and 1.4 - if they
have changed, you will have to change the pin definitions as well.

**Comment by HenrikPPersson** [comment]
Hi.
I also run with at SKR Pro 1.1 and the pin definition is the same as mine. As @dushyantahuja wrote, try switching over to Touch mode and then back again. That is what I'll have to do to make it works after a reset/reboot.

**Comment by SrGeek** [comment]
Any progress on this front?  I have a SKR mini E3 and only have TFT available to me.

**Comment by cata747** [comment]
Hi. Mine worked only using TFT35 folder NOT from unified menu material theme.  Did the mods from [https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/pull/705#issue-422238077](url) 
and switch back and forth from touch mode to marlin mode. 
I have skr 1.3 and TFT35 and in printer.cfg I have
[display]
lcd_type: st7920
cs_pin: P1.19
sclk_pin: P1.20
sid_pin: P1.18
encoder_pins: ^P3.26, ^P3.25
click_pin: ^!P0.28
I hope it helps someone.

**Comment by SrGeek** [comment]
@cata747 I only have one EXP1 and a TFT connection on my board.  There is no EXP2, thats where the encoder pins are at.

**Comment by SrGeek** [comment]
can I use a second mcu for my screen  i have an extra re-arm board it has exp1 & exp2.

**Comment by cata747** [comment]
@SrGeek There are config files for skr mini e3 for compiling the firmware. You can start there.
And some google [https://www.reddit.com/r/BIGTREETECH/comments/dtl8id/skr_mini_e3_12_klipper_configguide_including/](url)

**Comment by KevinOConnor** [comment]
Any further updates on this issue?  What's the current state of the BTT TFT on Klipper?

-Kevin

**Comment by trevjonez** [comment]
I've been running a custom build of the BTT firmware on my klipper machines to _some_ success.
As best I can tell the issue is purely on the BTT simulator logic because a real st7920 works fine. And their issue is still open and pinned for the char generator.

One nice thing is by doing a custom logo.bmp file I get a klipper logo for the splash screen on firmware restart. But the overall experience is far less than what has become consistent for klipper.

**Comment by dsbaars** [comment]
> Any further updates on this issue? What's the current state of the BTT TFT on Klipper?

Regarding **"Marlin Mode" / LCD12864 emulation mode**:
My TFT35 E3 V3.0 came with a firmware from June, this did not seem to work with Klipper. 
I compiled my own firmware, which works okay. Sometimes there are artefacts and sometimes you need to go to "touch mode" first and then to Marlin mode to make it work.
You can download my firmware and use it **at your own risk** from:
https://mega.nz/file/jbBUQKIa#xXKu3-K9MmVkF6lSxHjqni2rWLsxexGOXU1E_LxA-hc

I compiled this for an TFT35 **E3** V3.0 used with an Ender 3 (2020) with BTT SKR MINI E3 v2.0 firmware (note that there are two versions of the TFT35 V3.0, this one is for the fit-in replacement of the Ender 3 LCD, hence the "E3" addition).
 
The relevant part from `printer.cfg` to use this
````YAML
[board_pins]
aliases:
    # EXP1 header
    EXP1_1=PB5, EXP1_3=PA9,   EXP1_5=PA10, EXP1_7=PB8, EXP1_9=<GND>,
    EXP1_2=PA15, EXP1_4=<RST>, EXP1_6=PB9,  EXP1_8=PB15, EXP1_10=<5V>

[output_pin beeper]
pin: EXP1_1

[display]
lcd_type: st7920
cs_pin: EXP1_7
sclk_pin: EXP1_6
sid_pin: EXP1_8
encoder_pins: ^EXP1_5, ^EXP1_3
click_pin: ^!EXP1_2
````

While reading replies in several issues in the [BTT TFT Firmware repository](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/issues/281), I assume adding "read-only" support for "**Touchscreen mode**" should be quite simple at least the following two ways:
1) **IF** OctoPrint is able to "mirror" the data sent to Klipper to another serial port, connect your TFT directly to the GPIO pins of your Raspberry Pi (RX2/TX2/GND)
2) Re-send the commands which are sent from Octoprint to Klipper over the TFT interface of the SKR (E3) mainboards, for the [BTT SKR MINI E3 v2.0](https://github.com/bigtreetech/BIGTREETECH-SKR-mini-E3/blob/master/hardware/BTT%20SKR%20MINI%20E3%20V2.0/Hardware/BTT%20SKR%20MINI%20E3%20V2.0-PIN.pdf) this would be pins RX2/TX2

**Comment by alagerqvist** [comment]
Hi,

Sorry, I haven't tried anymore as I'am using marlin only.
I started on a emulator mode but then covid happened and shifted focus on
helping with visor prints etc.
/Anders

Den ons 26 feb. 2020 01:44DeviousPenguin <notifications@github.com> skrev:

**Comment by lilasiian8** [comment]
> > Any further updates on this issue? What's the current state of the BTT TFT on Klipper?
> 
> Regarding **"Marlin Mode" / LCD12864 emulation mode**:
> My TFT35 E3 V3.0 came with a firmware from June, this did not seem to work with Klipper.
> I compiled my own firmware, which works okay. Sometimes there are artefacts and sometimes you need to go to "touch mode" first and then to Marlin mode to make it work.
> You can download my firmware and use it **at your own risk** from:
> https://mega.nz/file/jbBUQKIa#xXKu3-K9MmVkF6lSxHjqni2rWLsxexGOXU1E_LxA-hc
> 
> I compiled this for an TFT35 **E3** V3.0 used with an Ender 3 (2020) with BTT SKR MINI E3 v2.0 firmware (note that there are two versions of the TFT35 V3.0, this one is for the fit-in replacement of the Ender 3 LCD, hence the "E3" addition).
> 
> The relevant part from `printer.cfg` to use this
> 
> ```yaml
> [board_pins]
> aliases:
>     # EXP1 header
>     EXP1_1=PB5, EXP1_3=PA9,   EXP1_5=PA10, EXP1_7=PB8, EXP1_9=<GND>,
>     EXP1_2=PA15, EXP1_4=<RST>, EXP1_6=PB9,  EXP1_8=PB15, EXP1_10=<5V>
> 
> [output_pin beeper]
> pin: EXP1_1
> 
> [display]
> lcd_type: st7920
> cs_pin: EXP1_7
> sclk_pin: EXP1_6
> sid_pin: EXP1_8
> encoder_pins: ^EXP1_5, ^EXP1_3
> click_pin: ^!EXP1_2
> ```
> 
> While reading replies in several issues in the [BTT TFT Firmware repository](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/issues/281), I assume adding "read-only" support for "**Touchscreen mode**" should be quite simple at least the following two ways:
> 
> 1. **IF** OctoPrint is able to "mirror" the data sent to Klipper to another serial port, connect your TFT directly to the GPIO pins of your Raspberry Pi (RX2/TX2/GND)
> 2. Re-send the commands which are sent from Octoprint to Klipper over the TFT interface of the SKR (E3) mainboards, for the [BTT SKR MINI E3 v2.0](https://github.com/bigtreetech/BIGTREETECH-SKR-mini-E3/blob/master/hardware/BTT%20SKR%20MINI%20E3%20V2.0/Hardware/BTT%20SKR%20MINI%20E3%20V2.0-PIN.pdf) this would be pins RX2/TX2

I can confirm that this successfully worked for me. Thank you for sharing this!

**Comment by artheoday** [comment]
How do I do this? Is there a tutorial to do this?


> > > Any further updates on this issue? What's the current state of the BTT TFT on Klipper?
> > 
> > 
> > Regarding **"Marlin Mode" / LCD12864 emulation mode**:
> > My TFT35 E3 V3.0 came with a firmware from June, this did not seem to work with Klipper.
> > I compiled my own firmware, which works okay. Sometimes there are artefacts and sometimes you need to go to "touch mode" first and then to Marlin mode to make it work.
> > You can download my firmware and use it **at your own risk** from:
> > https://mega.nz/file/jbBUQKIa#xXKu3-K9MmVkF6lSxHjqni2rWLsxexGOXU1E_LxA-hc
> > I compiled this for an TFT35 **E3** V3.0 used with an Ender 3 (2020) with BTT SKR MINI E3 v2.0 firmware (note that there are two versions of the TFT35 V3.0, this one is for the fit-in replacement of the Ender 3 LCD, hence the "E3" addition).
> > The relevant part from `printer.cfg` to use this
> > ```yaml
> > [board_pins]
> > aliases:
> >     # EXP1 header
> >     EXP1_1=PB5, EXP1_3=PA9,   EXP1_5=PA10, EXP1_7=PB8, EXP1_9=<GND>,
> >     EXP1_2=PA15, EXP1_4=<RST>, EXP1_6=PB9,  EXP1_8=PB15, EXP1_10=<5V>
> > 
> > [output_pin beeper]
> > pin: EXP1_1
> > 
> > [display]
> > lcd_type: st7920
> > cs_pin: EXP1_7
> > sclk_pin: EXP1_6
> > sid_pin: EXP1_8
> > encoder_pins: ^EXP1_5, ^EXP1_3
> > click_pin: ^!EXP1_2
> > ```
> > 
> > 
> > While reading replies in several issues in the [BTT TFT Firmware repository](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/issues/281), I assume adding "read-only" support for "**Touchscreen mode**" should be quite simple at least the following two ways:
> > 
> > 1. **IF** OctoPrint is able to "mirror" the data sent to Klipper to another serial port, connect your TFT directly to the GPIO pins of your Raspberry Pi (RX2/TX2/GND)
> > 2. Re-send the commands which are sent from Octoprint to Klipper over the TFT interface of the SKR (E3) mainboards, for the [BTT SKR MINI E3 v2.0](https://github.com/bigtreetech/BIGTREETECH-SKR-mini-E3/blob/master/hardware/BTT%20SKR%20MINI%20E3%20V2.0/Hardware/BTT%20SKR%20MINI%20E3%20V2.0-PIN.pdf) this would be pins RX2/TX2
> 
> I can confirm that this successfully worked for me. Thank you for sharing this!

**Comment by DroneMang** [comment]
The TFT24 display doesn't appear to work with Klipper.  I just tried this with the precompiled firmware from BTT and by compiling it, with the unified menu and standard, and nothing shows up other than LCD12864 simulator.  I have an SKR pro.

**Comment by derBG** [comment]
We are waiting for the source code and binary update by BTT

**Comment by basha318i** [comment]
Hello, 
I managed to get it working after changing SPI_ReEnable(!!(GPIOB->IDR & (1<<13))); to SPI_ReEnable(1)
but again still with some artefacts. this is how it looks on a BTT TFT35 V3.0 connected to an SKR 1.4 Turbo.
![IMG_7694](https://user-images.githubusercontent.com/555150/99309151-3a73ef00-2816-11eb-9876-4ba4c00833ae.JPG)
I'm not sure if it's just me or everyone is having this, but it looks really dull; like the font needs to be changed & made smaller; also I believe there's a better way of arranging stuff, similar to the Marlin Simulation, maybe?

**Comment by derBG** [comment]
Meep

**Comment by DroneMang** [comment]
I gave up and went with the stock reprap full display for now, if this
works that would be great! I got it working with the TFT 24 but I would
have to restart the firmware somewhere between 5 to 10 times to get the
display to actually come up.

**Comment by gkatev** [comment]
@github-actions Let's keep it open? :)

**Comment by MarsArtis** [comment]
> > Any further updates on this issue? What's the current state of the BTT TFT on Klipper?
> 
> Regarding **"Marlin Mode" / LCD12864 emulation mode**:
> My TFT35 E3 V3.0 came with a firmware from June, this did not seem to work with Klipper.
> I compiled my own firmware, which works okay. Sometimes there are artefacts and sometimes you need to go to "touch mode" first and then to Marlin mode to make it work.
> You can download my firmware and use it **at your own risk** from:
> https://mega.nz/file/jbBUQKIa#xXKu3-K9MmVkF6lSxHjqni2rWLsxexGOXU1E_LxA-hc
> 
> I compiled this for an TFT35 **E3** V3.0 used with an Ender 3 (2020) with BTT SKR MINI E3 v2.0 firmware (note that there are two versions of the TFT35 V3.0, this one is for the fit-in replacement of the Ender 3 LCD, hence the "E3" addition).
> 
> The relevant part from `printer.cfg` to use this
> 
> ```yaml
> [board_pins]
> aliases:
>     # EXP1 header
>     EXP1_1=PB5, EXP1_3=PA9,   EXP1_5=PA10, EXP1_7=PB8, EXP1_9=<GND>,
>     EXP1_2=PA15, EXP1_4=<RST>, EXP1_6=PB9,  EXP1_8=PB15, EXP1_10=<5V>
> 
> [output_pin beeper]
> pin: EXP1_1
> 
> [display]
> lcd_type: st7920
> cs_pin: EXP1_7
> sclk_pin: EXP1_6
> sid_pin: EXP1_8
> encoder_pins: ^EXP1_5, ^EXP1_3
> click_pin: ^!EXP1_2
> ```
> 
> While reading replies in several issues in the [BTT TFT Firmware repository](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/issues/281), I assume adding "read-only" support for "**Touchscreen mode**" should be quite simple at least the following two ways:
> 
> 1. **IF** OctoPrint is able to "mirror" the data sent to Klipper to another serial port, connect your TFT directly to the GPIO pins of your Raspberry Pi (RX2/TX2/GND)
> 2. Re-send the commands which are sent from Octoprint to Klipper over the TFT interface of the SKR (E3) mainboards, for the [BTT SKR MINI E3 v2.0](https://github.com/bigtreetech/BIGTREETECH-SKR-mini-E3/blob/master/hardware/BTT%20SKR%20MINI%20E3%20V2.0/Hardware/BTT%20SKR%20MINI%20E3%20V2.0-PIN.pdf) this would be pins RX2/TX2

Can you tell If this custom firmware may be used on TFT24 too, please?

**Comment by teeminus** [comment]
I just posted a reply to https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/issues/290: https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/issues/290#issuecomment-759543165

I think I fixed the issue, at least for me the TFT35v3 is now working with klipper. I also created a very stripped down and minimalistic version of the firmware that has just the ST7920 emulator left as I only use that mode.

**Comment by eremid** [comment]
Thanks @teeminus ! So if I want to test on my TFT35v3, i need the firmware in your fork repository and the config bellow ?

```
[display]
lcd_type: st7920
cs_pin: P1.19
sclk_pin: P1.20
sid_pin: P1.18
encoder_pins: ^P3.26, ^P3.25
click_pin: ^!P0.28
```

Thanks

**Comment by teeminus** [comment]
@eremid You can either use the firmware from the fork (you have to compile it yourself) or you use the precompiled binary from the stripped down firmware: https://github.com/teeminus/NoTouchScreenFirmware

I don't know if the klipper config is correct as I have a different mainboard. At least all required config entries are present, so if the pins are correct the display should work.

**Comment by MarsArtis** [comment]
> I just posted a reply to [bigtreetech/BIGTREETECH-TouchScreenFirmware#290](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/issues/290): [bigtreetech/BIGTREETECH-TouchScreenFirmware#290 (comment)](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/issues/290#issuecomment-759543165)
> 
> I think I fixed the issue, at least for me the TFT35v3 is now working with klipper. I also created a very stripped down and minimalistic version of the firmware that has just the ST7920 emulator left as I only use that mode.

Have tried your custom Firmware with my **BIGTREE_TFT24_V1_1**.
After the flashing is done I get this error:

**Illegal FLASH APP!**

**Comment by Sellecta** [comment]
I tried to use it on tft24, the screen is full of artifacts, maybe I got something wrong with the pins [https://photos.app.goo.gl/1nHccEQo3xTKoiyY6](url)

**Comment by dsbaars** [comment]
@MarsArtis @Sellecta the firmware I provided is for TFT35 (E3) only. Like @teeminus said, you have to compile it yourself.

**Comment by Sellecta** [comment]
> @MarsArtis @Sellecta the firmware I provided is for TFT35 (E3) only. Like @teeminus said, you have to compile it yourself.

@teeminus compiled for tft24 and I used it

**Comment by teeminus** [comment]
@MarsArtis Try compiling the firmware yourself. If that does not work either, check if the original BTT touch firmware works, I changed nothing on the compile stuff.

@Sellecta Someone else already reported that issue: https://github.com/teeminus/NoTouchScreenFirmware/issues/1
But I not quite sure what causes the problems. I am not familar with the STM32s and I don't have a TFT24 for debugging. As the code works in general (the TFT35 is running fine) I would guess the problem is causes by the TFT24 specific code (like clocks, timings, etc.).

**Comment by MarsArtis** [comment]
@teeminus

> @MarsArtis Try compiling the firmware yourself. If that does not work either, check if the original BTT touch firmware works, I changed nothing on the compile stuff.
not sure if the tools are available for OSX to compile, I'll try.
Anyway I did use the precompiled binary you make and wanted to report that it's not working.

**Comment by teeminus** [comment]
@MarsArtis You need platformio to compile the firmware. Edit the file platformio.ini and set default_envs=BIGTREE_TFT24_V1_1 to compile the firmware for the TFT24

**Comment by MarsArtis** [comment]
> @MarsArtis You need platformio to compile the firmware. Edit the file platformio.ini and set default_envs=BIGTREE_TFT24_V1_1 to compile the firmware for the TFT24

@teeminus 
Since I'm pretty new to PlatformIO I'm asking you for few tips:

1. how to "Disable fullscreen mode to prevent pixel overlap" ?
2. under **;Release mode** should I leave all the full string or just cut down to my TFT i.e. **default_envs = BIGTREE_TFT24_V1_1**?

**Comment by teeminus** [comment]
@MarsArtis I disabled the fullscreen mode in the last commit. Just checkout the master branch and you are good to go.

**Comment by MarsArtis** [comment]
> @MarsArtis I disabled the fullscreen mode in the last commit. Just checkout the master branch and you are good to go.

Ok, was not sure if I had to do it only in the config.ini or in the Platformio.ini itself. I did clone your git like 15 minutes ago.

**Comment by MarsArtis** [comment]
Some updates about TFT24 (**connected to a SKR 1.4 TURBO**)
Compiled from **teeminus** branch and now the display started without errors but with black screen and the text **LCD12864 Simulator**.
Have disabled fullscreen in config.ini like this:
```
#### Marlin Mode Fullscreen Support
# Run Marlin Mode in fullscreen.
#
# NOTE: Disable is recommended for TFT24.
#
#   Options: [enable: 1, disable: 0]
marlin_fullscreen:0

#### Marlin Mode Type
# Select Marlin Mode type.
#   Options: [LCD12864: 1, LCD2004: 0]
marlin_type:1
```
the klipper printer.cfg lines about the LCD :
```
[board_pins]
aliases:
    # EXP1 header
    EXP1_1=P1.30, EXP1_3=P1.18, EXP1_5=P1.20, EXP1_7=P1.22, EXP1_9=<GND>,
    EXP1_2=P0.28, EXP1_4=P1.19, EXP1_6=P1.21, EXP1_8=P1.23, EXP1_10=<5V>,
    # EXP2 header
    EXP2_1=P0.17, EXP2_3=P3.26, EXP2_5=P3.25, EXP2_7=P1.31, EXP2_9=<GND>,
    EXP2_2=P0.15, EXP2_4=P0.16, EXP2_6=P0.18, EXP2_8=<RST>, EXP2_10=<NC>

[display]
lcd_type: st7920 #BTT TFT24
cs_pin: EXP1_7
sclk_pin: EXP1_6
sid_pin: EXP1_8
encoder_pins: ^EXP1_5, ^EXP1_3
click_pin: ^!EXP1_2

[output_pin beeper]
pin: EXP1_1
```




any suggestion on next steps?

**Comment by teeminus** [comment]
@MarsArtis For my SKR Pro I have the following config (as suggested in the sample-lcd.cfg):
lcd_type: st7920
cs_pin: EXP1_4
sclk_pin: EXP1_5
sid_pin: EXP1_3
encoder_pins: ^EXP2_3, ^EXP2_5
click_pin: ^!EXP1_2

**Comment by MarsArtis** [comment]
> @MarsArtis For my SKR Pro I have the following config (as suggested in the sample-lcd.cfg):
> 
> [display]
> lcd_type: st7920
> cs_pin: EXP1_4
> sclk_pin: EXP1_5
> sid_pin: EXP1_3
> encoder_pins: ^EXP2_3, ^EXP2_5
> click_pin: ^!EXP1_2

have already tried out that, still the same black screen with text LCD12864 Simulator.
Btw, I've noticed that long pressing the knob on the TFT24 I'm still given the choice to select Marlin emulator and touch. Is that right?

To make sure I've compiled correctly the platformio.ini I'll paste here the initial strings:
```
[platformio]
src_dir      = src
boards_dir   = buildroot/boards
Build only Target = BIGTREE_TFT24_V1_1
default_envs = BIGTREE_TFT24_V1_1
```
and left everything else as is.

**Comment by MarsArtis** [comment]
Update:
I flashed by mistake the TFT28.bin on my TFT24.
So correctly using the TFT24 build from teeminus branch went "better".
Now the screen shows something more, but still garbled as many other people reported:

![IMG_20210115_233110](https://user-images.githubusercontent.com/49723675/104785184-3ae32480-578a-11eb-8111-de0aedb22f80.jpg)

**Comment by teeminus** [comment]
You could try the klipper st7920 driver modification that @HenrikPPersson mentioned above: https://github.com/KevinOConnor/klipper/issues/2231#issuecomment-632909375

**Comment by Sineos** [comment]
Just tried it for BIGTREE_TFT35_E3_V3_0 on SKR 1.4T

- Sometimes it works, sometimes it needs multiple firmware restarts to come up correctly
- Some elements like °C or the --> during heating are not displayed
- The menu comes up garbled but recovers once the selection is moved

@teeminus proposed code change improves this but is still kind of a hack and for me the screen starts freaking out once the HE fan starts.
A pity, but still not a fully working solution. Anyway, thanks for the efforts. Appreciated

**Comment by MarsArtis** [comment]
> You could try the klipper st7920 driver modification that @HenrikPPersson mentioned above: [#2231 (comment)](https://github.com/KevinOConnor/klipper/issues/2231#issuecomment-632909375)
> 
> EDIT: That code snippet does not work any longer due to changes in the klipper code. However, replacing the line `self.text_framebuffer[pos:pos+len(data)] = data` with
> 
> ```
>         for c in data:
>             self.write_graphics(x, y, font8x14.VGA_FONT[ord(c)])
>             x = x+1
> ```
> 
> works for me.

Well, I really don't know where to start about this. But If you want to instruct me I'll try.

**Comment by Sineos** [comment]
1. Open the file `st7920.py` in `klipper/klippy/extras/display` for editing
2. Find the line `self.text_framebuffer[pos:pos+len(data)] = data` (around line 137)
3. Replace the above with the pice of code @teeminus posted
4. Save the file and restart Klipper

**Comment by MarsArtis** [comment]
Realy thank @Sineos and @teeminus 
I do confirm that TFT24 is working as per your instructions.
At first (maybe very tired eyes) I did not find/saw the st7920.py in klipper/klippy/extras/display.

Modded and now it's showing clearly the screen text:

![TFT24](https://user-images.githubusercontent.com/49723675/104837673-66fbc400-58b6-11eb-9728-b89092174a91.png)
For testing purposes I warmed the nozzle to 50°C but the TFTscreen **remained at 21°** 


moreover The mainsail webgui is reporting me dirty Klipper version,I Suppose cause of the editing:
![Schermata 2021-01-17 alle 11 20 59](https://user-images.githubusercontent.com/49723675/104837686-87c41980-58b6-11eb-8438-dbfeb33a69b1.jpg)

**Comment by teeminus** [comment]
Glad to hear that the work around works. But I agree with @Sineos, the work around is quite hacky and I also experienced screen flicker during printing. So we need to figure out a better solution that does not require changing the klipper code.

**Comment by nicedevil007** [comment]
@teeminus

First, thank you for that "mod" :)

Unfurtunately I flashed your firmware on my BTT3.5 V3 (showed the flashing process)
Then I replaced the line in `klipper/klippy/extras/display/st7920.py` and die a restart of Klipper and the printer itself.

Screen stays black.

**Comment by MarsArtis** [comment]
@teeminus now I just need to make the LCD screen actually report whats going on with printer. As said it does not refresh temperatures

**Comment by mbgroot** [comment]
> I do confirm that TFT24 is working as per your instructions.
With what firmware?

**Comment by MarsArtis** [comment]
> > I do confirm that TFT24 is working as per your instructions.
> 
> With what firmware?

with the one provided by teeminus (you have to use his source and compile with PlatformIO). Follow the recent posts and you'll get how to

**Comment by CraXyOW3** [comment]
Awesome work @teeminus !
Just tried your source and just for the sake of it, one of the new touch themes on BTT's git.
Everything works as documented here, "->" sign and Celsius sign does not appear. As stated before, menu is a little garbled until you start scrolling.
For me, it looks like my content is a little more squished, like the width of the content does not stretch out on the sides.
But other than that, it works.

All in all, it feels that either BTT should fix better support for other firmwares than Marlin and/or we need awesome peeps that can maintain a custom firmware like @teeminus .

**Comment by teeminus** [comment]
I completely reimplemented the ST7920 emulator from scratch and updated my [NoTouchScreenFirmware](https://github.com/teeminus/NoTouchScreenFirmware).

The firmware now works without the need to modify the Klipper ST7920 driver. All screen artifacts are gone now, both the status screen and the menu work fine now (tested on the TFT35v3). Feel free to check that out and see if the firmware works for you as well.

**Comment by mbgroot** [comment]
On TFT24v11 black screen...

**Comment by MarsArtis** [comment]
> I completely reimplemented the ST7920 emulator from scratch and updated my [NoTouchScreenFirmware](https://github.com/teeminus/NoTouchScreenFirmware).
> 
> The firmware now works without the need to modify the Klipper ST7920 driver. All screen artifacts are gone now, both the status screen and the menu work fine now (tested on the TFT35v3). Feel free to check that out and see if the firmware works for you as well.

did a try right now against Klipper v0.9.1-205
flashed the rebuild .bin for TFT24 but all I get is a black screen with the text
NOTOUCHFW V1.1

**Comment by MarsArtis** [comment]
these are the first lines of the platformio.ini I've tuned for the build:

`; PlatformIO Project Configuration File
;
;   Build options: build flags, source filter
;   Upload options: custom upload port, speed and extra flags
;   Library options: dependencies, extra library storages
;   Advanced options: extra scripting
;
; Please visit documentation for the other options and examples
; https://docs.platformio.org/page/projectconf.html

;BIGTREE_TFT35_V3_0
;BIGTREE_TFT35_E3_V3_0
;BIGTREE_TFT35_B1_V3_0
;BIGTREE_TFT43_V3_0
;BIGTREE_TFT50_V3_0
;BIGTREE_TFT70_V3_0
;BIGTREE_TFT28_V3_0
;BIGTREE_TFT24_V1_1
;MKS_32_V1_4
;MKS_32_V1_4_NOBL
;MKS_28_V1_0

[platformio]
src_dir      = src
boards_dir   = buildroot/boards
Build only Target = BIGTREE_TFT24_V1_1
default_envs = BIGTREE_TFT24_V1_1

;Release mode
;default_envs = BIGTREE_TFT24_V1_1, BIGTREE_TFT28_V3_0, BIGTREE_TFT35_B1_V3_0, BIGTREE_TFT35_E3_V3_0, BIGTREE_TFT35_V3_0, BIGTREE_TFT43_V3_0, BIGTREE_TFT50_V3_0, BIGTREE_TFT70_V3_0, MKS_28_V1_0, MKS_32_V1_4, MKS_32_V1_4_NOBL
`

**Comment by teeminus** [comment]
@mbgroot @MarsArtis I created a new branch to debug the TFT24 issues where I reverted the SPI changes. The binary has been updated as well. Please check if that works.

The size of the binary has shrinked down to 18K as the complete touch functionality has been removed. There is only the Marlin mode left.

**Comment by MarsArtis** [comment]
@teeminus 
just downloaded **debug_TFT24** as suggested, compiled, flashed but still black screen with NOTOUCHFW V1.1 text 
![notiuch](https://user-images.githubusercontent.com/49723675/106724551-a0e3fe80-6608-11eb-8311-35e01993403f.jpg)


any other clue?

**Comment by teeminus** [comment]
Can you please have a look on the backside of the PCB which controller is used for the TFT24? Maybe there is a problem with the amount of RAM available.

Have you restarted klipper and the firmware after flashing the TFT?

**Comment by MarsArtis** [comment]
> Can you please have a look on the backside of the PCB which controller is used for the TFT24? Maybe there is a problem with the amount of RAM available.
> 
![btt24-1](https://user-images.githubusercontent.com/49723675/106726714-fde0b400-660a-11eb-814f-7acc6fb0ef9a.jpg)
![btt24-2](https://user-images.githubusercontent.com/49723675/106726720-00430e00-660b-11eb-8898-75f8930b8f4f.jpg)


> Have you restarted klipper and the firmware after flashing the TFT?
I did

**Comment by teeminus** [comment]
@MarsArtis  I recompiled the firmware using the original STM linker file. Maybe that helps. Could you please open a new ticket in the  [NoTouchScreenFirmware ](https://github.com/teeminus/NoTouchScreenFirmware) repo as the problem is now more related to the firmware than to klipper.

**Comment by MarsArtis** [comment]
@KevinOConnor 

we may say that everything is now working thanks to @teeminus custom firmware.
Without any Klipper hack he made a firmware that make the TFT24 works in Marlin mode withaout any glitch.

GREAT JOB

**Comment by jangrewe** [comment]
oooh, exciting! I'm a long-time Marlin user who converted his Artillery Sidewinder X1 to Klipper today, and the "idle" display really bothered me. Marlin Mode by itself is better than a touchscreen that's showing only wrong info.

AFAIK the SWX1 has a MKS TFT28 3.2 in there, i'm going to try your FW as soon as i'm done with all the Klipper tuning instructions.

**Comment by sagreen83** [comment]
Is anyone actively working on adding TFT35 support to Klipper?

**Comment by moebis** [comment]
Too bad it's not possible to bypass the marlin mode completely and built a nice color and touch interface for Klipper and TFT35

**Comment by Jhyrachy** [comment]
in theory is possible, the problem is the knowledge to do it

**Comment by Vybramo** [comment]
> Realy thank @Sineos and @teeminus
> I do confirm that TFT24 is working as per your instructions.
> At first (maybe very tired eyes) I did not find/saw the st7920.py in klipper/klippy/extras/display.
> 
> Modded and now it's showing clearly the screen text:
> 
> ![TFT24](https://user-images.githubusercontent.com/49723675/104837673-66fbc400-58b6-11eb-9728-b89092174a91.png)
> For testing purposes I warmed the nozzle to 50°C but the TFTscreen **remained at 21°**
> 
> moreover The mainsail webgui is reporting me dirty Klipper version,I Suppose cause of the editing:
> ![Schermata 2021-01-17 alle 11 20 59](https://user-images.githubusercontent.com/49723675/104837686-87c41980-58b6-11eb-8438-dbfeb33a69b1.jpg)

How you let ur display work ? Its will be good for me just use it in marlin mode.

**Comment by SmartFabian** [comment]
Hello,

the mod is working for me, but I have the same flicker during printing as @teeminus.
Only during printing 80% of the display turns black for 1 sec and then show again everything for 1 sec during the hole print. Is there a workaround?

Greetings from Germany

**Comment by pixeldoc2000** [comment]
@teeminus 

Did a Test on my **BIQU B1** with the stock "BIGTREE TFT35 V3.0 B1 Touchscreen" flashed with your NoTouchFW v1.3.
It work fine. Thank you for your awesome work!

![grafik](https://user-images.githubusercontent.com/376715/123527923-cd95d100-d6e3-11eb-8ce7-a8df3330cbab.png)

fluidd: v1.15.0 
klipper: v0.9.1-586-g8bce8924 
moonraker: v0.6.0-160-gd3df568

printer.cfg
`[display]
lcd_type: st7920
cs_pin: EXP1_4
sclk_pin: EXP1_5
sid_pin: EXP1_3
encoder_pins: ^EXP2_3, ^EXP2_5
click_pin: ^!EXP1_2`

**Comment by ihrapsa** [comment]
> Is anyone actively working on adding TFT35 support (native BTT touch firmware) to Klipper?

If anyone's interested in adding TFT touchscreen support for Klipper, check Desuuuu's work on the [runtime-uart](https://github.com/Desuuuu/klipper/commits/runtime-uart) branch. He's been working on a sort of serial pass-through implementation that will allow dynamically control of TFTs directly through mainboard's secondary serial interface. Maybe someone could use that to finally port the btt touchscreen TFT in Klipper 😄

**Comment by atomadog** [comment]
> @teeminus
> 
> Did a Test on my **BIQU B1** with the stock "BIGTREE TFT35 V3.0 B1 Touchscreen" flashed with your NoTouchFW v1.3 .
> 
> ![grafik](https://user-images.githubusercontent.com/376715/123527923-cd95d100-d6e3-11eb-8ce7-a8df3330cbab.png)
> 
> **It works fine. Thank you for your awesome work!**
> 
> fluidd: v1.15.0 klipper: v0.9.1-586-g8bce8924 moonraker: v0.6.0-160-gd3df568
> 
> printer.cfg
> 
> ```
> [display]
> lcd_type: st7920
> cs_pin: EXP1_4
> sclk_pin: EXP1_5
> sid_pin: EXP1_3
> encoder_pins: ^EXP2_3, ^EXP2_5
> click_pin: ^!EXP1_2
> ```
> 
> **EDIT:** The Display does not react to Input and does not Update after starting a print.

Use this

[display]
lcd_type: emulated_st7920
spi_software_miso_pin: EXP2_1
spi_software_mosi_pin: EXP1_3
spi_software_sclk_pin: EXP1_5
en_pin: EXP1_4
encoder_pins: ^EXP2_5, ^EXP2_3
click_pin: ^!EXP1_2

**Comment by pixeldoc2000** [comment]
@atomadog 

Thanks atomadog, but I have isssus with Both `lcd_type: st7920` and `lcd_type: emulated_st7920` on my BIQU B1 as the Display freezes at random after starting a Print.
See https://github.com/teeminus/NoTouchScreenFirmware/issues/47 and https://github.com/teeminus/NoTouchScreenFirmware/discussions/43

**Comment by hapklaar** [comment]
@pixeldoc2000 

Currently I have the screen on the B1 working almost without issue (as in freezes or screen corruption). I use v 1.3.1 and the following config in printer.cfg.

```
[display]
lcd_type: emulated_st7920
spi_software_miso_pin: EXP2_1
spi_software_mosi_pin: EXP1_3
spi_software_sclk_pin: EXP1_5
spi_speed: 1000000
en_pin: EXP1_4
encoder_pins: ^EXP2_3, ^EXP2_5
click_pin: ^!EXP1_2
```

**Comment by pixeldoc2000** [comment]
@hapklaar 

You are using a different `spi_software_miso_pin: EXP2_1`. I guess this is not by accident? How did you discover it? 

With v1.3.1 you mean [NoTouchFirmware](https://github.com/teeminus/NoTouchScreenFirmware) i guess? 

 I will give it a try and Report back.

**Comment by hapklaar** [comment]
@pixeldoc2000 Can't remember where I got the miso_pin config. I do indeed mean the NoTouchFirmware 1.3.1

**Comment by teeminus** [comment]
There has been a discussion of on how to choose the MISO here: https://github.com/teeminus/NoTouchScreenFirmware/issues/25

**Comment by pixeldoc2000** [comment]
@hapklaar 
Still the same issus with FW 1.3.1 and different pins.

Starting to think it is an issus with my Printer or my Hardware. What Hardware do you have connected?
I have an original BLTouch 3.1 and the Stock Filament Runout Sensor connected to the Mainboard instead of the Display.

**Comment by hapklaar** [comment]
> @hapklaar Still the same issus with FW 1.3.1 and different pins.
> 
> Starting to think it is an issus with my Printer or my Hardware. What Hardware do you have connected? I have an original BLTouch 3.1 and the Stock Filament Runout Sensor connected to the Mainboard instead of the Display.

Also original BLTouch 3.1, filament runout sensor still connected to display. I replaced the tmc2208's at some point with 2209's. I run all the latest klipper software and firmware (updated with kiauh)

**Comment by pixeldoc2000** [comment]
Tried [github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware ](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware) in Marlin Mode. I get the same issue as with [NoTouchFirmware](https://github.com/teeminus/NoTouchScreenFirmware), so looks like it is definitely a problem with my setup/hardware.

**Comment by mlaci001** [comment]
The same is true with the E3 turbo and TFT35-E3-v3.
The screen is blank.
If the TFT button is held down for 3 seconds, the mode selection window will appear, here you have to select the Marlin mode (no matter what mode the screen is in).
The screen will still be blank, but if you press the button again, the menu will appear and work from there.
You have to do it again every time you start.
I don't know why it's not good at startup.
I can't use emulated_st7920 mode, the motherboard only has an EXP1 connector.

**Comment by mlaci001** [comment]
@pixeldoc2000
[Original firmware]:
The screen is blank.
If the TFT button is held down for 3 seconds, the mode selection window will appear, here you have to select the Marlin mode (no matter what mode the screen is in).
The screen will still be blank, but if you press the button again, the menu will appear and work from there.
[NoTouchFirmware]:
Cold boot ok, but after firmware restart (klipper) garbage characters come up.

**Comment by r00t4rd3d** [comment]
This is kinda unbelievable that the TFT35 and Klipper cant work together in touch screen mode. I dont get why the devs dont fix this themselves? BTT has their own version of Klipper so why the hell dont they fix this crap?

**Comment by lucasromeiro** [comment]
Any new about this. I have the same problem 😞

**Comment by miracuruzrh** [comment]
Hello

I had the same issue like others with the BTT TFT50v3 display. I use it together with the BTT SKR v1.4 turbo mainboard.
The issue was the same like the first screenshot with the cryptic lines and dots. What helped for me is to update the firmware of the TFT using the [git firmware repository](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware). After the update it suddenly worked.

Here is the printer.cfg section which works for me:
```
[board_pins]
aliases:
    # EXP1 header
    EXP1_1=P1.30, EXP1_3=P1.18, EXP1_5=P1.20, EXP1_7=P1.22, EXP1_9=<GND>, # XXX
    EXP1_2=P0.28, EXP1_4=P1.19, EXP1_6=P1.21, EXP1_8=P1.23, EXP1_10=<5V>, # XXX
    # EXP2 header
    EXP2_1=P0.17, EXP2_3=P3.26, EXP2_5=P3.25, EXP2_7=P1.31, EXP2_9=<GND>, # XXX
    EXP2_2=P0.15, EXP2_4=P0.16, EXP2_6=P0.18, EXP2_8=<RST>, EXP2_10=<NC>  # XXX
    # Pins EXP2_1, EXP2_6, EXP2_2 are also MISO, MOSI, SCK of bus "ssp0"
    
    [display]
lcd_type: emulated_st7920
encoder_pins: ^EXP2_3, ^EXP2_5
spi_software_miso_pin: EXP2_1
spi_software_mosi_pin: EXP1_3
spi_software_sclk_pin: EXP1_5
en_pin: EXP1_4
encoder_pins: ^EXP2_5, ^EXP2_3
click_pin: ^!EXP1_2
spi_speed: 1000000

# #display_group:
menu_timeout: 1
```

**Comment by Minims** [comment]
@KevinOConnor Is it possible to run Klipper on this screen using a stm32 chip (I think yes) and use it like in `[mcu]` as display like the Voron V0 display does ? Maybe the issue is on `lcd_type` here ? Thanks.
